ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,765 @@
+///
+/// Corresponds to `AccumulatorState` of the MAC based malicious validator.
+#[derive(Clone, Debug)]
+struct Batch {
+    chunk_size: usize,
+    inner: HashMap<Gate, UnverifiedValuesStore>,
+}
+
+impl Batch {
+    fn new(chunk_size: usize) -> Self {
+        Self {
+            chunk_size,
+            inner: HashMap::<Gate, UnverifiedValuesStore>::new(),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.inner.is_empty() || self.inner.values().all(UnverifiedValuesStore::is_empty)
+    }
+
+    fn push(&mut self, gate: Gate, record_id: RecordId, segment: Segment) {
+        // get value_store & create new when necessary
+        // insert segment
+        self.inner
+            .entry(gate)
+            .or_insert_with(|| UnverifiedValuesStore::new(self.chunk_size))
+            .insert_segment(record_id, segment);
+    }
+
+    /// This function should only be called by `validate`!
+    ///
+    /// Updates all `UnverifiedValueStores` in hashmap by incrementing the offset to next chunk
+    /// and deallocating `vec`.
+    ///
+    /// ## Panics
+    /// Panics when `UnverifiedValuesStore` panics, i.e. when `segment_size` is `None`
+    fn update(&mut self) {
+        self.inner
+            .values_mut()
+            .for_each(UnverifiedValuesStore::update);
+    }
+
+    /// `get_unverified_field_value` converts a `Batch` into an iterator over `UnverifiedFieldValues`
+    /// compatible with DZKPs
+    #[allow(dead_code)]
+    fn get_unverified_field_values<DF: DZKPBaseField>(
+        &self,
+    ) -> impl Iterator<Item = DF::UnverifiedFieldValues> + '_ {
+        self.inner
+            .values()
+            .flat_map(UnverifiedValuesStore::get_unverified_field_values::<DF>)
+    }
+}
+
+/// Corresponds to `MaliciousAccumulator` of the MAC based malicious validator.
+#[derive(Clone, Debug)]
+pub struct DZKPBatch {
+    inner: Weak<Mutex<Batch>>,
+}
+
+impl DZKPBatch {
+    /// pushes values of a record, i.e. segment, to a `Batch`
+    ///
+    /// ## Panics
+    /// Panics when mutex is poisoned or `segments` have different lengths within `gate`
+    pub fn push(&self, gate: Gate, record_id: RecordId, segment: Segment) {
+        let arc_mutex = self.inner.upgrade().unwrap();
+        // LOCK BEGIN
+        let mut batch = arc_mutex.lock().unwrap();
+        batch.push(gate, record_id, segment);
+        // LOCK END
+    }
+
+    /// ## Panics
+    /// Panics when mutex is poisoned
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        // return true if there is either no value (upgrade returns none) or there is a value but is_empty() is true
+        !self
+            .inner
+            .upgrade()
+            .is_some_and(|x| !x.lock().unwrap().is_empty())
+    }
+}
+
+/// Steps used by the validation component of the DZKP
+#[cfg_attr(feature = "descriptive-gate", derive(ipa_macros::Step))]
+pub(crate) enum Step {
+    /// For the execution of the malicious protocol.
+    DZKPMaliciousProtocol,
```

Comment:
lets strip the prefix, we know that everything that happens in this module is related to zkp validation

Reply:
This could cause a collision with MACs since the same context might be upgraded to MACs and DZKPs. It will probably call different subroutines with narrow. I still think it is a bit safer to have a different name.